src/commands.rs

```diff
@@ -3216,28 +3236,54 @@ fn show_predecessor_patch(
     diff_format: DiffFormat,
 ) -> Result<(), CommandError> {
     if let Some(predecessor) = commit.predecessors().first() {
-        let predecessor_tree = if predecessor.parent_ids() == commit.parent_ids() {
-            predecessor.tree()
-        } else {
-            // Rebase the predecessor to have the current commit's parent(s) and use that
-            // tree as base
-            let new_parent_tree =
-                merge_commit_trees(workspace_command.repo().as_repo_ref(), &commit.parents());
-            let old_parent_tree = merge_commit_trees(
-                workspace_command.repo().as_repo_ref(),
-                &predecessor.parents(),
-            );
-            let rebased_tree_id =
-                merge_trees(&new_parent_tree, &old_parent_tree, &predecessor.tree())?;
-            workspace_command
-                .repo()
-                .store()
-                .get_tree(&RepoPath::root(), &rebased_tree_id)?
-        };
+        let predecessor_tree = rebase_to_other_parent(workspace_command, predecessor, commit)?;
         let diff_iterator = predecessor_tree.diff(&commit.tree(), &EverythingMatcher);
-        show_diff(formatter, workspace_command, diff_iterator, diff_format)?;
+        show_diff(formatter, workspace_command, diff_iterator, diff_format)
+    } else {
+        Ok(())
```

Comment:
Not that I know of :) You could move the `Ok()` out of the if/else if you add `?;` after the `show_diff()` call. I don't know which would be considered more idiomatic. It's probably just a matter of personal preference (I'm fine with either form).

Reply:
You can use this pattern, which would scale for multiple early returns:

```rust
    let predecessors = commit.predecessors();
    let predecessor = match predecessors.first() {
        Some(predecessor) => predecessor,
        None => return Ok(()),
    };

    let predecessor_tree = rebase_to_other_parent(workspace_command, predecessor, commit)?;
    let diff_iterator = predecessor_tree.diff(&commit.tree(), &EverythingMatcher);
    show_diff(formatter, workspace_command, diff_iterator, diff_format)
```

Note that `predecessors` has to be extracted into a local variable because otherwise the borrowed `predecessor` would escape the scope of the `match`, which is inconvenient.

It's up to you if you prefer to write it this way; it might be clearer as-is.

Reply:
> Note that predecessors has to be extracted into a local variable because otherwise the borrowed predecessor would escape the scope of the match, which is inconvenient.

Interesting! TIL. I assume this is an example of [place expression vs match expression(http://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/reference/expressions/match-expr.html)?

i.e. `commit.predecessors().first()` is a value expression (because of the `.predecessors()`) and would get dropped at the end of the match, whereas `predecessors.first()` is a place expression and doesn't get dropped.

If so, that's a pretty subtle distinction, hm.